src/jdk.incubator.vector/share/classes/jdk/incubator/vector/AbstractVector.java

```diff
@@ -188,12 +188,53 @@ ByteVector asByteVectorRawTemplate() {
 
     abstract AbstractMask<E> maskFromArray(boolean[] bits);
 
-    abstract AbstractShuffle<E> iotaShuffle();
+    abstract <F> VectorShuffle<F> toShuffle(AbstractSpecies<F> dsp);
 
-    abstract AbstractShuffle<E> iotaShuffle(int start, int step, boolean wrap);
+    /*package-private*/
+    @ForceInline
+    final <F> VectorShuffle<F> toShuffleTemplate(AbstractSpecies<F> dsp) {
+        Class<?> etype = vspecies().elementType();
+        Class<?> dvtype = dsp.shuffleType();
+        Class<?> dtype = dsp.asIntegral().elementType();
+        int dlength = dsp.dummyVector().length();
+        return VectorSupport.convert(VectorSupport.VECTOR_OP_CAST,
+                                     getClass(), etype, length(),
+                                     dvtype, dtype, dlength,
+                                     this, dsp,
+                                     AbstractVector::toShuffle0);
+    }
+
+    abstract <F> VectorShuffle<F> toShuffle0(AbstractSpecies<F> dsp);
+
+    @ForceInline
+    public final
+    VectorShuffle<E> toShuffle() {
+        return toShuffle(vspecies());
+    }
 
-    /*do not alias this byte array*/
-    abstract AbstractShuffle<E> shuffleFromBytes(byte[] reorder);
+    abstract VectorShuffle<E> iotaShuffle();
+
+    @ForceInline
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    final VectorShuffle<E> iotaShuffle(int start, int step, boolean wrap) {
+        if ((length() & (length() - 1)) != 0) {
+            return wrap ? shuffleFromOp(i -> (VectorIntrinsics.wrapToRange(i * step + start, length())))
+                        : shuffleFromOp(i -> i * step + start);
+        }
+
+        AbstractSpecies<?> species = vspecies().asIntegral();
+        Vector iota = species.iota();
```

Comment:
Thanks a lot for your review, I think that transforming a multiplication by a power of 2 into a shift can be done by the C2 compiler. I have added the special case for `start = 0 && step == 1` since it may be more common and can be optimised away when the arguments are constants.

Reply:
For x86 byte vector multiplication is done at granularity of short lanes, this case shows regression with power of two multiplications which are strength reduced to shifts currently. please file a follow up bug report for this.

Reply:
@jatin-bhateja I have created [JDK-8305810](https://bugs.openjdk.org/browse/JDK-8305810)